### DISCUSSION TITLE
node.data TS Errors

### DISCUSSION THREAD

**Discussion by hmaiermack** [xyflow/xyflow#2959] [Q&A]
Hi, running into a TS error when updating a node. Using the UpdateNode example:

```js
setNodes((nds) => {
   nds.map((node) => {
       if (node.id === '1') {
          node.data = {
               //this throws a TS error
               //Error: Unsafe assignment of an `any` value.  @typescript-eslint/no- 
               //unsafe-assignment
               ...node.data, <- error
               label: nodeName
         }
       }
      ....
``` 
I've gotten past the error via: `...node.data as {[key: string]: unknown}`
but am wondering if there's a better way to go about typing this.

**Comment by moklick** [comment]
Hey @hmaiermack ,

can you share the rest of the code? It's possible to pass a type as a generic to define the shape of `data`.

```ts
type NodeData = {
  some: 'foo'
};

const nodes: Node<NodeData>[] = [ ... ]
```

Does it help?

**Comment by hmaiermack** [comment]
Taken from the workflow builder example, I'm doing this in a couple of places:
```ts
      const currNodes: Node[] = transitions.map(({ node}) => {
        return {
          id: node.id,
          position: {},
          data: { ...node.data as unknown as { [key: string]: unknown}},
          type: node.type,
        };
      });
```
In my project a node can be of n types each containing different metadata. I could maintain types for data for each different node type but I'd expect inference in this case. I'm importing my nodes and edges into an index file and then exporting them like so:
```ts
import type { EdgeTypes, NodeTypes } from "reactflow";
//node and edge imports follow

const nodeTypes: NodeTypes = {
   //my node types
}

const edgeTypes: EdgeTypes = {
   //my edge types
}

export { nodeTypes, edgeTypes }
``` 
Is it possible I'm doing something wrong? I disabled the eslint rule so it's not an issue anymore, just curious if I should be going about this differently. If the answer is to maintain type information for the data object of each possible node type I'll probably stick with my workaround.
